src/pnpq/devices/switch_thorlabs_osw_e.py

```diff
@@ -213,32 +224,103 @@ def get_state(self) -> State:
-            return response.decode("utf-8")
+            try:
+                return response.decode("utf-8")
+            except UnicodeDecodeError as e:
+                raise ThorlabsOswError(
+                    code=None,
+                    description="Received non-decodable type code from Thorlabs OSW device.",
+                    raw_reply=response.decode("utf-8", errors="replace"),
+                ) from e
 
     def get_board_name(self) -> str:
+        """
+        Get the switch name and firmware version.
+        :return: human-readable, unstructured string.
+        :raises: ThorlabsOswError: If the device reports an error or sends an unexpected reply.
+        """
         self._fail_if_closed()
         with self._communication_lock:
             command = b"I?\n"
             self._connection.write(command)
             response = self._read_serial_response()
-            return response.decode("utf-8")
+            try:
+                return response.decode("utf-8")
+            except UnicodeDecodeError as e:
+                raise ThorlabsOswError(
+                    code=None,
+                    description="Received non-decodable board name from Thorlabs OSW device.",
+                    raw_reply=response.decode("utf-8", errors="replace"),
+                ) from e
 
     def _read_serial_response(self) -> bytes:
-        """Read a response from the serial connection."""
-        response = self._connection.read_until(b"\r\n")[:-2]  # Remove the trailing \r\n
-        return response
+        """Read a response from the serial connection.
+
+        All error that report from the device will translate into 'ThorlabsOswError' exceptions.
+        """
+        try:
+            response = self._connection.read_until(b"\r\n")
+        except Exception as e:
+            raise ThorlabsOswError(
+                code=None,
+                description="Failed to read response from Thorlabs OSW device.",
+                raw_reply="",
+            ) from e
+
+        if not response:
+            raise ThorlabsOswError(
+                code=None,
+                description="No response received from Thorlabs OSW device.",
+                raw_reply="",
+            )
+        if response.endswith(b"\r\n"):
+            decoding_messages = response[:-2]
+        else:
+            decoding_messages = response
+
+        # Decode to inspect for "Error.." messages
+        try:
+            text = decoding_messages.decode("utf-8")
+        except UnicodeDecodeError as e:
+            raise ThorlabsOswError(
+                code=None,
+                description="Received non-decodable response from Thorlabs OSW device.",
+                raw_reply=response.decode("utf-8", errors="replace"),
+            ) from e
+        stripped_text = text.strip()
+        if stripped_text.lower().startswith("error"):
+            raise parse_thorlabs_osw_error(stripped_text)
+        return decoding_messages
```

Comment:
If we're already decoding to a string... shouldn't we just return the string here?